staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -0,0 +1,617 @@
+		return fmt.Sprintf("%v/%v:%v", s.Namespace, s.Name, s.Port)
+	}
+	return s.LocalName
+}
+
+func newServiceKey(service apiregistrationv1.ServiceReference) serviceKey {
+	// Docs say. Defaults to 443 for compatibility reasons.
+	// BETA: Should this be a shared constant to avoid drifting with the
+	// implementation?
+	port := int32(443)
+	if service.Port != nil {
+		port = *service.Port
+	}
+
+	return serviceKey{
+		Name:      service.Name,
+		Namespace: service.Namespace,
+		Port:      port,
+	}
+}
+
+type cachedResult struct {
+	// Currently cached discovery document for this service
+	// Map from group name to version name to
+	discovery map[metav1.GroupVersion]apidiscoveryv2beta1.APIVersionDiscovery
+
+	// ETag hash of the cached discoveryDocument
+	etag string
+
+	// Guaranteed to be a time less than the time the server responded with the
+	// discovery data.
+	lastUpdated time.Time
+
+	// results are stale if server was attempted to be contacted and it failed
+	// if lastUpdated is beyond a certain threshold, and stale is true, the result
+	// will be purged
+	stale bool
+}
+
+// Information about a specific APIService/GroupVersion
+type groupVersionInfo struct {
+	// Date this APIService was marked dirty.
+	// Guaranteed to be a time greater than the most recent time the APIService
+	// was known to be modified.
+	//
+	// Used for request deduplication to ensure the data used to reconcile each
+	// apiservice was retrieved after the time of the APIService change:
+	// real_apiservice_change_time < groupVersionInfo.lastMarkedDirty < cachedResult.lastUpdated < real_document_fresh_time
+	//
+	// This ensures that if the apiservice was changed after the last cached entry
+	// was stored, the discovery document will always be re-fetched.
+	lastMarkedDirty time.Time
+
+	// Last time sync function was run for this GV.
+	lastReconciled time.Time
+
+	// ServiceReference of this GroupVersion. This identifies the Service which
+	// describes how to contact the server responsible for this GroupVersion.
+	service serviceKey
+
+	// Method for contacting the service
+	handler http.Handler
+}
+
+var _ DiscoveryAggregationController = &discoveryManager{}
+
+func NewDiscoveryManager(
+	target discoveryendpoint.ResourceManager,
+) DiscoveryAggregationController {
+	return &discoveryManager{
+		mergedDiscoveryHandler: target,
+		apiServices:            make(map[string]groupVersionInfo),
+		cachedResults:          make(map[serviceKey]cachedResult),
+		localDelegates:         make(map[string]genericapiserver.DelegationTarget),
+		dirtyAPIServiceQueue:   workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "discovery-manager"),
+	}
+}
+
+// Returns discovery data for the given apiservice.
+// Caches the result.
+// Returns the cached result if it is retrieved after the apiservice was last
+// marked dirty
+func (dm *discoveryManager) fetchFreshDiscoveryForService(gv metav1.GroupVersion, info groupVersionInfo) (cachedResult, error) {
```

Comment:
Let's either
1. switch the return to be *cachedResult, error so we can return a nil when we fail
2. return an explicit zero value for ever call site returning an error.  

As I read it now, I can be returning a stale cached result from deep in this function.